packages/eslint-plugin/src/lib/util.ts

```diff
@@ -0,0 +1,144 @@
+import { Rule } from 'eslint';
+import { ParserServices } from '@typescript-eslint/typescript-estree';
+
+const version = require('../../package.json').version;
+
+export type ObjectLike<T = any> = Record<string, T>;
+
+export function tslintRule(name: string) {
+  return `\`${name}\` from TSLint`;
+}
+
+export function metaDocsUrl(name: string) {
+  return `https://github.com/typescript-eslint/typescript-eslint/blob/${version}/packages/eslint-plugin/docs/rules/${name}.md`;
+}
+
+/**
+ * Check if the context file name is *.ts or *.tsx
+ * @param fileName The context file name
+ * @returns `true` if the file name ends in *.ts or *.tsx
+ * @private
+ */
+export function isTypescript(fileName: string) {
+  return /\.tsx?$/i.test(fileName || '');
+}
+
+/**
+ * Check if the context file name is *.d.ts or *.d.ts
+ * @param fileName The context file name
+ * @returns `true` if the file name ends in *.d.ts or *.d.ts
+ * @private
+ */
+export function isDefinitionFile(fileName: string) {
+  return /\.d\.tsx?$/i.test(fileName || '');
+}
+
+/**
+ * Check if the variable contains an object stricly rejecting arrays
+ * @param obj an object
+ * @returns `true` if obj is an object
+ */
+function isObjectNotArray<T extends object>(obj: T | any[]): obj is T {
+  return typeof obj === 'object' && !Array.isArray(obj);
+}
+
+/**
+ * Pure function - doesn't mutate either parameter!
+ * Merges two objects together deeply, overwriting the properties in first with the properties in second
+ * @param first The first object
+ * @param second The second object
+ * @returns a new object
+ */
+export function deepMerge<T extends ObjectLike = ObjectLike>(
+  first: ObjectLike = {},
+  second: ObjectLike = {}
+): T {
+  // get the unique set of keys across both objects
+  const keys = new Set(Object.keys(first).concat(Object.keys(second)));
+
+  return Array.from(keys).reduce(
+    (acc, key) => {
+      const firstHasKey = key in first;
+      const secondHasKey = key in second;
+
+      if (firstHasKey && secondHasKey) {
+        if (isObjectNotArray(first[key]) && isObjectNotArray(second[key])) {
+          // object type
+          acc[key] = deepMerge(first[key], second[key]);
+        } else {
+          // value type
+          acc[key] = second[key];
+        }
+      } else if (firstHasKey) {
+        acc[key] = first[key];
+      } else {
+        acc[key] = second[key];
+      }
+
+      return acc;
+    },
+    {} as T
+  );
+}
+
+/**
+ * Pure function - doesn't mutate either parameter!
+ * Uses the default options and overrides with the options provided by the user
+ * @template TOptions
+ * @param defaultOptions the defaults
+ * @param userOptions the user opts
+ * @returns the options with defaults
+ */
+export function applyDefault<T extends any[]>(
```

Comment:
can we use `ObjectLike` here instead of `any`? we shouldn't ever be trying to apply defaults to a non-object

Reply:
In this case, no.
eslint options are not required to be objects - they can be any valid json value.

I.e. you could have a set of string options for your rule: `["error", "spaces", 2]` (`"error"` will be stripped from the options, giving you just the last 2).
Defaults for this option could thus be `["spaces", 4]`, which then gets merged together with the user provided options.

If the user provides no options `["error"]`, then we are given an empty array, and we'll fill in the blanks with the default elements, returning `["spaces", 4]`.

This method takes an extra step as well in that if it detects an object, it'll do a deep merge of the default object, and the user provided object.